File: src/lib/java/work/lclpnet/ap2/impl/base/PlayerManagerImpl.java
package work.lclpnet.ap2.impl.base;

import net.fabricmc.fabric.api.networking.v1.PlayerLookup;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.level.ServerPlayer;
import work.lclpnet.ap2.api.base.ParticipantListener;
import work.lclpnet.ap2.api.base.PlayerManager;

import java.util.*;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;
import java.util.stream.Collectors;

public class PlayerManagerImpl implements PlayerManager {

    private final MinecraftServer server;
    private final Set<UUID> participants = new HashSet<>();
    private final Set<UUID> permanentSpectators = new HashSet<>();
    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
    private final Lock readLock = readWriteLock.readLock();
    private final Lock writeLock = readWriteLock.writeLock();
    private boolean prepare = false;
    private boolean finale = false;
    private ParticipantListener listener = null;

    public PlayerManagerImpl(MinecraftServer server) {
        this.server = server;
    }

    @Override
    public Set<ServerPlayer> getAsSet() {
        try {
            readLock.lock();

            return PlayerLookup.all(server).stream()
                    .filter(player -> participants.contains(player.getUUID()))
                    .collect(Collectors.toUnmodifiableSet());
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public boolean isParticipating(UUID uuid) {
        Objects.requireNonNull(uuid);

        try {
            readLock.lock();

            return participants.contains(uuid);
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public boolean offer(ServerPlayer player) {
        Objects.requireNonNull(player);

        try {
            readLock.lock();

            if (!prepare || finale) return false;
        } finally {
            readLock.unlock();
        }

        try {
            writeLock.lock();

            participants.add(player.getUUID());
            return true;
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void startPreparation() {
        setPreparation(true);
        reset();
    }

    @Override
    public void startMiniGame() {
        setPreparation(false);
        reset();
    }

    private void setPreparation(boolean prepare) {
        try {
            writeLock.lock();

            this.prepare = prepare;
        } finally {
            writeLock.unlock();
        }
    }

    private void reset() {
        try {
            writeLock.lock();

            if (finale) {
                // remove finalists who left
                participants.removeIf(uuid -> server.getPlayerList().getPlayer(uuid) == null);
            } else {
                addAllPlayers();
            }
        } finally {
            writeLock.unlock();
        }
    }

    private void addAllPlayers() {
        // clear old entries in case the player has left
        participants.clear();

        PlayerLookup.all(server).stream()
                .map(ServerPlayer::getUUID)
                .filter(uuid -> !permanentSpectators.contains(uuid))
                .forEach(participants::add);
    }

    @Override
    public void enterFinale(Set<? extends ServerPlayer> finalists) {
        try {
            writeLock.lock();

            finale = true;
            participants.clear();

            finalists.stream()
                    .map(ServerPlayer::getUUID)
                    .forEach(participants::add);
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public boolean isPermanentSpectator(ServerPlayer player) {
        try {
            readLock.lock();

            return permanentSpectators.contains(player.getUUID());
        } finally {
            readLock.unlock();
        }
    }

    @Override
    public void addPermanentSpectator(ServerPlayer player) {
        try {
            writeLock.lock();

            permanentSpectators.add(player.getUUID());
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void removePermanentSpectator(ServerPlayer player) {
        try {
            writeLock.lock();

            permanentSpectators.remove(player.getUUID());
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void remove(ServerPlayer player) {
        try {
            writeLock.lock();

            if (participants.remove(player.getUUID()) && listener != null) {
                listener.participantRemoved(player);
            }
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public void bind(ParticipantListener listener) {
        this.listener = listener;
    }

    @Override
    public Optional<ServerPlayer> getParticipant(UUID uuid) {
        if (!participants.contains(uuid)) {
            return Optional.empty();
        }

        return Optional.ofNullable(server.getPlayerList().getPlayer(uuid));
    }

    @Override
    public void leaveFinale() {
        try {
            writeLock.lock();

            finale = false;

            addAllPlayers();
        } finally {
            writeLock.unlock();
        }
    }

    @Override
    public boolean isFinale() {
        return finale;
    }
}

File: src/lib/src/main/kotlin/work/lclpnet/ap2/game/player/PlayerManagerImpl.kt
package work.lclpnet.ap2.game.player

import net.fabricmc.fabric.api.networking.v1.PlayerLookup
import net.minecraft.server.MinecraftServer
import net.minecraft.server.level.ServerPlayer
import work.lclpnet.ap2.game.data.type.PlayerRef
import java.util.*

class PlayerManagerImpl(private val server: MinecraftServer) : PlayerManager {

    private val lock = Any()
    private val participants = HashSet<UUID>()
    private val permanentSpectators = HashSet<UUID>()
    private val initial = HashSet<PlayerRef>()
    private var prepare = false
    private var finale = false
    private var listener: ParticipantListener? = null

    override val asSet: Set<ServerPlayer> get() = synchronized(lock) {
        PlayerLookup.all(server)
            .filter { it.uuid in participants }
            .toSet()
    }

    override fun isParticipating(uuid: UUID): Boolean = synchronized(lock) {
        uuid in participants
    }

    override fun offer(player: ServerPlayer): Boolean = synchronized(lock) {
        if (!prepare || finale) return@synchronized false

        participants.add(player.uuid)
        true
    }

    override fun startPreparation(): Unit = synchronized(lock) {
        prepare = true
        reset()
    }

    override fun startMiniGame(): Unit = synchronized(lock) {
        prepare = false
        reset()
        setInitial()
    }

    private fun setInitial() {
        for (uuid in participants) {
            val player = server.playerList.getPlayer(uuid)
            initial.add(if (player != null) PlayerRef.create(player) else PlayerRef.createForUuid(uuid))
        }
    }

    private fun reset() {
        initial.clear()

        if (finale) {
            // remove finalists who left
            participants.removeAll { server.playerList.getPlayer(it) == null }
        } else {
            addAllPlayers()
        }
    }

    private fun addAllPlayers() {
        // clear old entries in case the player has left
        participants.clear()

        PlayerLookup.all(server)
            .map { it.uuid }
            .filter { it !in permanentSpectators }
            .forEach { participants.add(it) }
    }

    override fun enterFinale(finalists: Set<ServerPlayer>): Unit = synchronized(lock) {
        finale = true
        participants.clear()

        finalists.forEach { participants.add(it.uuid) }
    }

    override fun isPermanentSpectator(player: ServerPlayer): Boolean = synchronized(lock) {
        player.uuid in permanentSpectators
    }

    override fun addPermanentSpectator(player: ServerPlayer): Unit = synchronized(lock) {
        permanentSpectators.add(player.uuid)

        if (prepare && !finale && participants.remove(player.uuid)) {
            listener?.participantRemoved(player)
        }
    }

    override fun removePermanentSpectator(player: ServerPlayer): Unit = synchronized(lock) {
        permanentSpectators.remove(player.uuid)

        if (prepare && !finale) {
            participants.add(player.uuid)
        }
    }

    override fun remove(player: ServerPlayer): Unit = synchronized(lock) {
        if (participants.remove(player.uuid)) {
            listener?.participantRemoved(player)
        }
    }

    override fun bind(listener: ParticipantListener?): Unit = synchronized(lock) {
        this.listener = listener
    }

    override fun getParticipant(uuid: UUID): ServerPlayer? = synchronized(lock) {
        if (uuid !in participants) {
            null
        } else {
            server.playerList.getPlayer(uuid)
        }
    }

    override fun leaveFinale(): Unit = synchronized(lock) {
        finale = false
        addAllPlayers()
    }

    override val isFinale: Boolean get() = synchronized(lock) {
        finale
    }

    override val initialParticipants: Set<PlayerRef> get() = synchronized(lock) {
        initial.toSet()
    }
}